### PR TITLE
Sync default stream in `DeviceBuffer` constructor

### PR DESCRIPTION
**Pull request by jakirkham**
To make sure data is completely copied over to `DeviceBuffer` when using the default stream, make sure to sync on it after constructing the `rmm::device_buffer` object. This ensures it is allocated and its data (if any is provided) is fully populated.

### CODE REVIEW

**Comment by kkraus14** [comment]
I would expect this to be handled in the C++ as opposed to the Python. @jrhemstad what is your perspective here?

**Comment by jrhemstad** [comment]
> I would expect this to be handled in the C++ as opposed to the Python. @jrhemstad what is your perspective here?

Definitely not. `device_buffer` provides no synchronization whatsoever. That would defeat the purpose of it being stream ordered. 

I'm concerned by how much the Python side is synchronizing the default stream. That is going to be very expensive and totally kill us when/if we ever expose and use streams in libcudf.

**Comment by kkraus14** [comment]
Understood, @jakirkham what's the situation that is biting us here? Sending data via UCX-py?

**Comment by jakirkham** [comment]
There have been some issues with accuracy. I suspect (though it's a bit difficult to prove) this is happening due to copies not completing by the time a comparison occurs. So thinking syncing here would help. As we are already doing this in other APIs here, figured this was a pretty reasonable addition.

**Comment by harrism** [comment]
So sync before the comparison, rather than inside the DeviceBuffer code?

**Comment by kkraus14** [comment]
> There have been some issues with accuracy. I suspect (though it's a bit difficult to prove) this is happening due to copies not completing by the time a comparison occurs. So thinking syncing here would help. As we are already doing this in other APIs here, figured this was a pretty reasonable addition.

It sounds like we should be synchronizing in UCX-py before transmitting data. I agree it makes sense to synchronize here for now at least, but we'll likely want to explore how we can enable asynchronous behavior on the default stream to enable better usage of default stream per thread which we'll want to enable in the future.

**Comment by jakirkham** [comment]
So I think it is useful to see this in context.

```python
hb = np.arange(5, dtype="uint8")

ptr = hb.__array_interface__["data"][0]
(size,) = hb.__array_interface__["shape"]

db = rmm.DeviceBuffer(ptr=ptr, size=size)
# Do something with `db` (what happens here?)
```

Python users generally expect that `db` is fully initialized once they use it. Currently that is not the case. This fixes that. I think this expectation of behavior is independent of whether it is used in ucx-py, Dask, cuDF, or other downstream user code.

More generally we seem to be having an ongoing conversation about performance and user expectations. This seems to come up a lot around syncing in particular. As we aim to interoperate with the broader community, I think the latter is more important than the former (when we have to choose between them) as it determines the experience people have when they drop in our libraries in their workflows. Though this may just be my personal bias 🙂

**Comment by jrhemstad** [comment]
> More generally we seem to be having an ongoing conversation about performance and user expectations. This seems to come up a lot around syncing in particular. As we aim to interoperate with the broader community, I think the latter is more important than the former (when we have to choose between them) as it determines the experience people have when they drop in our libraries in their workflows.

I think this is correct. The sticking point for me is that it doesn't seem like there's an option to opt out of the synchronization and enable power users.

**Comment by jakirkham** [comment]
Right now the user can control this by specifying a non-default stream, which is what we have done so far. Does that satisfy this need sufficiently? If not, what else do you think would be worth exposing here?

**Comment by jrhemstad** [comment]
> Right now the user can control this by specifying a non-default stream, which is what we have done so far. Does that satisfy this need sufficiently? If not, what else do you think would be worth exposing here?

I would think a person should be able to construct a `DeviceBuffer` on the default stream without needing a synchronization.

**Comment by jakirkham** [comment]
Why is that needed?

**Comment by kkraus14** [comment]
> Why is that needed?

Because there's a compiler option of `per-thread default stream` which allows you to program on the default stream as you normally would and then get nice kernel execution overlap by using multiple host threads. With our current model they'd still want to create objects on the "default" stream, but there's no way to do so asynchronously which presents a problem to them.

**Comment by kkraus14** [comment]
Also even without that compiler option someone could want to copy to the device asynchronously, run a few kernels, then get the results, and the best performance option would be to synchronize only at the end, but we don't allow doing that on the default stream currently.

**Comment by jrhemstad** [comment]
> Because there's a compiler option of per-thread default stream which allows you to program on the default stream as you normally would and then get nice kernel execution overlap by using multiple host threads. With our current model they'd still want to create objects on the "default" stream, but there's no way to do so asynchronously which presents a problem to them.

This is definitely not the only reason why you want to avoid syncing the default stream. 

Syncing the default stream has the same effect as a `cudaDeviceSynchronize()`. It brings the entire GPU to a halt synchronizing _all_ other streams.

**Comment by kkraus14** [comment]
@jrhemstad super noob question, but is there a good way to synchronize only the default stream without synchronizing the whole device?

**Comment by jrhemstad** [comment]
> @jrhemstad super noob question, but is there a good way to synchronize only the default stream without synchronizing the whole device?

Only if you've compiled in per-thread default stream mode.

In non-PTDS mode, the default stream is special because it implicitly synchronizes. Meaning:
```
kernel1<<<..., stream1>>>(...);
cudaStreamSynchronize(0); // sync default stream
kernel3<<<..., stream2>>>(...);
```

`kernel1` and `kernel3` can never overlap because of the implicit synchronization caused by the stream synchronize on the default stream. 

PTDS has two side-effects:
- Each thread's default stream is independent
- The default stream is a "regular stream" and isn't ordered wrt to other streams.
   - This means in the above example that kernel1, the stream sync, and kernel3 can all overlap.

**Comment by kkraus14** [comment]
That's what I thought and is very unfortunate. Unfortunately the way CUDA Python libraries have gone thus far is that if you don't specify a stream everything is synchronous, otherwise you go full on asynchronous with manual explicit synchronization. I think we should continue with this PR as is and then revisit if needed when we move to per thread default stream.

File: python/rmm/_lib/device_buffer.pyx [hunk 1]
```diff
@@ -46,6 +46,7 @@ cdef class DeviceBuffer:
                   uintptr_t stream=0):
```
Comment (kkraus14) [inline python/rmm/_lib/device_buffer.pyx]: @jakirkham we should add documentation to the init function here to clarify the synchronization behavior with regards to streams here.

**Comment by jakirkham** [comment]
It's worth noting this is something we are already doing throughout this code. Here's [one example]( https://github.com/rapidsai/rmm/blob/38a2be78ccad0395f46bd8574305be324b6b54df/python/rmm/_lib/device_buffer.pyx#L442-L446 ). Though one can find multiple usages in this code. At this point not syncing on the default stream in our code is an exception not the norm. So it seems reasonable to align to it here and match the existing behavior that is relied upon.

Should add this will also fix an issue in `DeviceBuffer` that we have around `__cuda_array_interface__` being synchronous. ( https://github.com/numba/numba/issues/4933 )

File: python/rmm/_lib/device_buffer.pyx (hunk 1 above)
Comment (jakirkham) [reply python/rmm/_lib/device_buffer.pyx]: Agreed. Thanks for mentioning. Have added docs for the constructor along with this note.

Comment (jakirkham) [reply python/rmm/_lib/device_buffer.pyx]: Also double checked that we have this note in other places where we synchronize.

**Comment by jrhemstad** [comment]
> It's worth noting this is something we are already doing throughout this code. Here's [one example](https://github.com/rapidsai/rmm/blob/38a2be78ccad0395f46bd8574305be324b6b54df/python/rmm/_lib/device_buffer.pyx#L442-L446). Though one can find multiple usages in this code. At this point not syncing on the default stream in our code is an exception not the norm.

Yeah, that's what I'm afraid of :(

**Comment by jakirkham** [comment]
Could you please raise an issue @jrhemstad? 🙂

**Comment by jakirkham** [comment]
> Could you please raise an issue @jrhemstad? 🙂

This was raised as issue ( https://github.com/rapidsai/rmm/issues/313 ).

**Comment by jakirkham** [comment]
This has put us in closer compliance with synchronization requirements of `__cuda_array_interface__` ( https://github.com/numba/numba/pull/5162 ). Though we may still need a change to ensure non-default streams are handled correctly.

**Comment by leofang** [comment]
I am confused. Isn’t `cudaMalloc` a synchronous function? Why bother doing extra sync?

**Comment by jrhemstad** [comment]
> I am confused. Isn’t `cudaMalloc` a synchronous function? Why bother doing extra sync?

RMM has options for using an asynchronous, stream ordered pool sub-allocator via [cnmem](https://github.com/NVIDIA/cnmem) and other experimental implementations. 

Furthermore, one of the `device_buffer` ctors copies from a host/device buffer using `cudaMemcpyAsync` on the specified stream.

**Comment by jakirkham** [comment]
It was primarily the copy that we needed to sync for.

**Comment by leofang** [comment]
Ah, OK, thanks Jake & John. Didn't know about cnmem...Is it turned on in RMM by default?

**Comment by jakirkham** [comment]
Of course. 🙂 Yeah it is

**Comment by jrhemstad** [comment]
@leofang @jakirkham 

Cnmem is not enabled by default in the c++ library. I didn't think we were enabling it by default in Python either.

**Comment by jakirkham** [comment]
My understanding of the question was we always build CNMeM support. Is that not correct?

**Comment by jrhemstad** [comment]
> My understanding of the question was we always build CNMeM support. Is that not correct?

It's always compiled, yes, but not used unless initialized in pool mode.

**Pull request merged**